File: python/coin_change.py
class Solution:
    def coinChange(self, coins: list[int], amount: int) -> int:
        dp: list[int] = [amount + 1] * (amount + 1)
        dp[0] = 0

        for a in range(1, amount + 1):
            for c in coins:
                if a - c >= 0:
                    dp[a] = min(dp[a], 1 + dp[a - c])

        return dp[amount] if dp[amount] != amount + 1 else -1


def test_coinChange_case_1():
    # arrange
    coins: list[int] = [1, 2, 5]
    amount: int = 11
    expected: int = 3

    # act
    solution = Solution()
    result = solution.coinChange(coins, amount)

    # assert
    assert result == expected


def test_coinChange_case_2():
    # arrange
    coins: list[int] = [2]
    amount: int = 3
    expected: int = -1

    # act
    solution = Solution()
    result = solution.coinChange(coins, amount)

    # assert
    assert result == expected


def test_coinChange_case_3():
    # arrange
    coins: list[int] = [1]
    amount: int = 0
    expected: int = 0

    # act
    solution = Solution()
    result = solution.coinChange(coins, amount)

    # assert
    assert result == expected

File: rust/src/code/coin_change.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity O(n * m) where n is amount and m is coins length
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let mut dp: Vec<i32> = vec![amount + 1; amount as usize + 1];
        dp[0] = 0;

        for i in 1..(amount + 1) {
            for c in coins.iter() {
                if i - c >= 0 {
                    dp[i as usize] = std::cmp::min(dp[i as usize], 1 + dp[(i - c) as usize]);
                }
            }
        }

        return if dp[amount as usize] == amount + 1 {
            -1
        } else {
            dp[amount as usize]
        };
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_coin_change_case_1() {
        // arrange
        let coins = vec![1, 2, 5];
        let amount = 11;
        let expected = 3;

        // act
        let result = Solution::coin_change(coins, amount);

        // assert
        assert_eq!(result, expected);
    }

    #[test]
    fn test_coin_change_case_2() {
        // arrange
        let coins = vec![2];
        let amount = 3;
        let expected = -1;

        // act
        let result = Solution::coin_change(coins, amount);

        // assert
        assert_eq!(result, expected);
    }

    #[test]
    fn test_coin_change_case_3() {
        // arrange
        let coins = vec![1];
        let amount = 0;
        let expected = 0;

        // act
        let result = Solution::coin_change(coins, amount);

        // assert
        assert_eq!(result, expected);
    }
}
